### ISSUE TITLE
Magnetic trait deletes items when the effect is active while using a magnet from another mod.

### ISSUE THREAD

**Issue by Awri-Lynn** [SlimeKnights/TinkersConstruct#3857]
While playing Skyfactory 4  (4.0.3)  I was wearing a boots with magnetic trait (iron trim) and while equipped, The effects of attracting items by Fluxomagnet from CoFH, or Magnet from PickeTweaks causes them to void instead.
From my tests, it only happens with Tinkers Magnetic trait, when it is active by an armor OR tool use with it, it deletes the item completely voids it.
From my googling, similar item related bugs by Tinkers Construct caused deleting or duplicating, so I suspect there is something weird with TC's item handling.

**Didn't crash**

**Versions:**
* Minecraft: 1.12.2
* Forge: 14.23.5.2831
* Mantle: 1.12-1.3.3.49
* Tinkers Construct: TConstruct-1.12.2-2.12.0.135
(Playing modpack: SkyFactory 4.0.3)

**Tinkers Construct and a magnet from other mods**

**Comment by KnightMiner** [comment]
I find it very unlikely that Tinkers Construct is causing this bug, all we do is [add velocity to the item](https://github.com/SlimeKnights/TinkersConstruct/blob/1.12/src/main/java/slimeknights/tconstruct/tools/traits/TraitMagnetic.java#L88-L90). Are you sure the item is deleted and not just sent flying away from the player? Try testing in a creative world using `/say @e[type=item]` to verify if the item still exists

**Comment by Awri-Lynn** [comment]
Well, they said it's not their magnet, you say it's not your magnet effect, but it is the two of you together that voids, actually, it's tinker's magnet effect that does it to any other magnet.

My current solution is just to avoid tinker's with iron parts from now on.

**Comment by KnightMiner** [comment]
Can you link me to the issue where "they" said its not their magent? It would help to talk to the other devs directly rather than both of us saying "not our fault"

**Comment by Awri-Lynn** [comment]
I tried to approach Team CoFH through their Discord server, and ask casually in their bug-report chat.
The King Lemming himself said in chat "There's something xmod going on. Probably another mod with an item event that isn't handling cancels right."

Forgot to mention that I'll be doing the testing in my next game session, I just re-read my last comment and it sounded angry, sorry, Poe's law in effect here.

If you guys have a discord, have a link somewhere, I could not find one.

**Comment by KnightMiner** [comment]
There is a link to our discord [at the top of the CurseForge a page](https://minecraft.curseforge.com/projects/tinkers-construct)

**Comment by KnightMiner** [comment]
@Awri-Lynn can you give us more information on what is happening that causes the magnets together to void the items? I have been talking to King Lemming and he agrees there is no reason it should void between the two of them, neither of us do anything beyond setting the item's position and velocity, so apart from a really obscure vanilla bug its nearly impossible to tell whats going on.

Best guess is you have something else in your inventory that voids the item, for example on of the Thermal Expansion Satchels. Please get a minimal setup to reproduce this so we can verify the cause of the issue.

**Comment by krisavi** [comment]
Having same issue in Engineer's Life mod pack. Over there the Pickle is used. As far as what I have seen, then I can use many of other mods magnets and items will not disappear. But when the Magnetic trait comes to play, then items start to disappear. I have Tinker's Armory used.

I will test with only Tinker's Tools.

So issue seems to be more on Armory side not TCon Tools. 
https://gyazo.com/e6e9c2296f2ed71387263d58c540e4fe

**Comment by KnightMiner** [comment]
If its on the armor side, thats from an addon mod, however the logic is the same. We just add velocity to the item, there is no reason adding velocity would delete the item unless another magnet is doing something super hacky.

You have a satchel in your inventory, have you checked to see if it is grabbing the items?

**Comment by krisavi** [comment]
yeap, t is on taskbar, but it is empty, this version of Iron Packbacks does not even support autocollection. 
I checked the code yes, seemed quite the same, except one break command, but that should be there to avoid pulling too many items per tick. 
The magnet in PickleTweaks just sets the location to player location. But for some reason having those mods together ends up eating the items.

**Comment by bonii-xx** [comment]
We normalize the vector and apply a constant force when doing the pull. So even if the "teleportation" by their magnet does weird stuff, worst case the item ends marginally behind you. I really doubt this is something that's fixable on our side. If we ever find the cause.

**Comment by krisavi** [comment]
I was thinking the same, that maybe it just moves really fast as one mod teleports item. Not sure if teleport is considered as movement of item. That would already make the velocity of the item extremely fast (3-4 blocks in one tick) and adding extra force could mean that it just flies away in opposite way, but that does not seem to be the case. I tried to encase myself in chunk size cubicle, dropped the item in the center below me, so the teleportation would be minimal and also force added would be minimal and it should not "fly" away somewhere. The item still disappears. I am wondering if a mod that makes the item to be laying on ground instead of hovering like in vanilla could be piece why in combination items disappear. I have not tested with only the 3-4 mods setup. All the tests I have done are using the modpack. Checking the code of the magnet in PickleTweaks and modifying it so it would not pull the items off ImmersiveEngineering conveyor belts that have scaffolding and in progress adding some debug information did not give me any answers either.

Could it for some reason put ItemEntity into a block and in a way kill it? I am not that familiar with the mechanics of ItemEntity and collision, so not sure if that is causing the item.setDead().

**Comment by KnightMiner** [comment]
Items in blocks just get shot upwards, they don't die. The technical Minecraft community has used it for item elevators before.

**Comment by KnightMiner** [comment]
So, but late, but someone on Discord found the cause. As I predicted, the issue is not the item being voided, the item is getting an unusual velocity causing it to move to an invalid position. Would have probably found earlier if people ever answered the first question I asked on this issue instead of just saying others say its not their fault.

Specifically, our magnet code calculates the direction from the item to the player. If the item coordinates ever exactly equal the players (impossible with vanilla, possible with people who do "magnets" by teleporting), that vector is not properly normalized as Java's VecMath library does not follow the convention of 0 vector normalizing to 0.

The code I wrote for 1.16 does not have this problem as Mojang's new vector library stuff correctly handles the 0 vector. For 1.12, depends on if we get another build.
